From be8bd838963cda1e8171213bb87091f8a2a962c8 Mon Sep 17 00:00:00 2001
From: Chaitanya Kulkarni <chaitanya.kulkarni@wdc.com>
Date: Thu, 1 Oct 2020 12:37:50 -0700
Subject: [PATCH 4/4] nvmet: allow user to mark file-ns write-protected

Right now we open file-ns in read-write mode. This doesn't allow us to
export the file which is marked read-only.

Allow user to export the readonly file by considering the
ns->write_protect configfs attribute when opening the file. Use O_RDWR
when ns->write_protect is not set.

Signed-off-by: Chaitanya Kulkarni <chaitanya.kulkarni@wdc.com>
---
 drivers/nvme/target/io-cmd-file.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/nvme/target/io-cmd-file.c b/drivers/nvme/target/io-cmd-file.c
index 7fdbdc496597..c096602d1207 100644
--- a/drivers/nvme/target/io-cmd-file.c
+++ b/drivers/nvme/target/io-cmd-file.c
@@ -41,9 +41,21 @@ void nvmet_file_ns_disable(struct nvmet_ns *ns)
 
 int nvmet_file_ns_enable(struct nvmet_ns *ns)
 {
-	int flags = O_RDWR | O_LARGEFILE;
+	int flags = O_LARGEFILE;
 	int ret;
 
+	switch (ns->write_protect) {
+	case NVME_NS_WRITE_PROTECT_PERMANENT:
+	case NVME_NS_WRITE_PROTECT:
+		flags |= O_RDONLY;
+		break;
+	case NVME_NS_NO_WRITE_PROTECT:
+		flags |= O_RDONLY | O_RDWR;
+		break;
+	default:
+		return -EINVAL;
+	}
+
 	if (!ns->buffered_io)
 		flags |= O_DIRECT;
 
-- 
2.22.1

